File: dart/binary_search_tree.dart
class Node {
  String data;
  Node? left;
  Node? right;
  Node(this.data);
}

class BinarySearchTree {
  Node root;

  BinarySearchTree(String data) : root = Node(data);

  void insert(String data) {
    root = _insert(data, root);
  }

  static Node _insert(String data, Node? node) {
    if (node == null) {
      return Node(data);
    }

    if (data.compareTo(node.data) <= 0) {
      node.left = _insert(data, node.left);
    } else {
      node.right = _insert(data, node.right);
    }
    return node;
  }

  List<String> get sortedData {
    List<String> arr = [];
    List<Node> st = [];
    Node? curr = root;
    while (true) {
      if (curr != null) {
        st.add(curr);
        curr = curr.left;
      } else {
        if (st.isEmpty) {
          break;
        }
        curr = st.removeLast();
        arr.add(curr.data);
        curr = curr.right;
      }
    }
    return arr;
  }
}

File: go/binary_search_tree.go
package binarysearchtree

type BinarySearchTree struct {
	left  *BinarySearchTree
	data  int
	right *BinarySearchTree
}

// NewBst creates and returns a new BinarySearchTree.
func NewBst(i int) *BinarySearchTree {
	return &BinarySearchTree{
		data: i,
	}
}

func insert(bst *BinarySearchTree, i int) *BinarySearchTree {
	if bst == nil {
		return NewBst(i)
	}

	if bst.data < i {
		bst.right = insert(bst.right, i)
	} else {
		bst.left = insert(bst.left, i)
	}
	return bst
}

// Insert inserts an int into the BinarySearchTree.
// Inserts happen based on the rules of a binary search tree
func (bst *BinarySearchTree) Insert(i int) {
	bst = insert(bst, i)
}

// SortedData returns the ordered contents of BinarySearchTree as an []int.
// The values are in increasing order starting with the lowest int value.
// A BinarySearchTree that has the numbers [1,3,7,5] added will return the
// []int [1,3,5,7].
func (bst *BinarySearchTree) SortedData() []int {
	curr := bst
	st := []*BinarySearchTree{}
	arr := []int{}
	for curr != nil || len(st) > 0 {
		for curr != nil {
			st = append(st, curr)
			curr = curr.left
		}

		curr = st[len(st)-1]
		arr = append(arr, curr.data)
		st = st[:len(st)-1]
		curr = curr.right
	}
	return arr
}
